### ISSUE TITLE
postConceptSetDefinition is still not working and I cannot find correct DEV branch with functioning code.

### ISSUE THREAD

**Issue by BradleySlavikAstellas** [OHDSI/ROhdsiWebApi#234]
Sample code?

The usual code to connect to ATLAS via ROhdsiWebApi

Retrieve a concept set definition, then try to use it to create a copy of it.

conceptSet ← getConceptSetDefinition(conceptSetId = 518, baseUrl = webapi_root)
View(conceptSet)
postConceptSetDefinition(name = “BS Test concept”, conceptSetDefinition = conceptSet, baseUrl = webapi_root)
Here is the error, which might help shed light on the issue

Error in .request(url, method = “PUT”, config = config, handle = handle, : Bad Request (HTTP 400).

Is there a permission problem? Is our ATLAS instance misconfigured?
I have been using R programs to post cohort definitions and that code is working well.

Thanks for any suggestions on what I should do differently.

Bradley Slavik

P.S. Issue #224 closed on Oct 20 and claimed that code was in 1.2.1, I have tried 1.2.1 several times. It does not work. Same error.

**Comment by azimov** [comment]
Hi Bradley, the API is a little confusing when used in this context, it looks like the mistake that you're making is that the object `conceptSet` returned by `getConceptSetDefinition` is actually not the expression you want to send when posting a new definition.

`postConceptSetDefinition` actually only accepts the expression part of the `conceptSet` and not the other meta-data such as its creator, its name (which need to be set when calling the post function) and unique id.

The following code should work:

```
conceptSet ← getConceptSetDefinition(conceptSetId = 518, baseUrl = webapi_root)
postConceptSetDefinition(name = “BS Test concept”, **conceptSetDefinition = conceptSet$expression**, baseUrl = webapi_root)
```

**Comment by BradleySlavikAstellas** [comment]
Hello Jamie,

Yeah, that was my bad in typing that. Did you try the code? Adding $expression changes nothing. It still does not work. I just reloaded version 1.2.1 again to be sure that I was using the correct version.

Bradley Slavik
(847)208-7194

**Comment by azimov** [comment]
Hi Bradley, I can't confirm your issue on any of the atlas instances I have access to. Are you correctly logged in? 
It's possible to have an account that has read access but not write access. Use `ROhdsiWebApi::authorizeWebApi` to log in.

When you re-installed `ROhdsiWebApi` did you restart your R session? If not the old version of the package may be loaded.

Also, from what you posted, It's not possible to see what your concept set expression actually contains. It could also be related to the expression itself; if it uses a format or vocabulary that was accepted in an old version of ATLAS when you try and upload the new version the validation fails. 

I'm also unable to test on ATLAS version `2.7.8` so I can't tell if it is a bug that as been fixed in a newer webapi.

**Comment by BradleySlavikAstellas** [comment]
ConceptSetDefinition gets created. There are no concepts in it. Same code **was** working a few months ago when I tested it. Yes you are correct that I needed to pass in R object with $items under it. All that checks out. Something is going wrong with the code to add concepts under conceptSetDefinition. Perhaps it is an ATLAS issue. I don't have access to log files for WebAPI. Is there a log file I can ask Odyssey (who provides support for our instance), to look at?

Thanks!
